### PR TITLE
mds: add more debug logs before crashing the MDS

### PR DESCRIPTION
**Pull request by lxbsz**
Print the detail conflicting requests before crashing the MDS.

Fixes: https://tracker.ceph.com/issues/57044
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by ajarr — APPROVED** [review]
The code looks good. I'm not sure about the commit message saying that it fixes https://tracker.ceph.com/issues/56633, which is the issue of duplicate/conflicting request ID.

**Comment by lxbsz** [comment]
> The code looks good. I'm not sure about the commit message saying that it fixes https://tracker.ceph.com/issues/56633, which is the issue of duplicate/conflicting request ID.

Not a fix, just a test commit for this issue. Maybe `s/Fixes: URL:` ?

File: src/mds/MDCache.cc [hunk 1]
```diff
@@ -9652,9 +9652,14 @@ MDRequestRef MDCache::request_start_internal(int op)
   MDRequestRef mdr =
       mds->op_tracker.create_request<MDRequestImpl,MDRequestImpl::Params*>(&params);
 
-  ceph_assert(active_requests.count(mdr->reqid) == 0);
+  if (active_requests.count(mdr->reqid)) {
+    auto _mdr = active_requests[mdr->reqid];
+    dout(0) << __func__ << " existing " << *_mdr << " op " << _mdr->internal_op << dendl;
+    dout(0) << __func__ << " new " << *mdr << " op " << op << dendl;
+    ceph_assert(active_requests.count(mdr->reqid) == 0);
```
Comment (vshankar) [inline src/mds/MDCache.cc:9659]: very minor nit: could use ceph_abort() since the details have been logged anyway.

Comment (lxbsz) [reply src/mds/MDCache.cc:9659]: Sure, will fix it.

**Review by dparmar18 — COMMENTED** [review]
Looks good

File: src/mds/MDCache.cc [hunk 2]
```diff
@@ -9652,9 +9652,14 @@ MDRequestRef MDCache::request_start_internal(int op)
   MDRequestRef mdr =
       mds->op_tracker.create_request<MDRequestImpl,MDRequestImpl::Params*>(&params);
 
-  ceph_assert(active_requests.count(mdr->reqid) == 0);
+  if (active_requests.count(mdr->reqid)) {
+    auto _mdr = active_requests[mdr->reqid];
```
Comment (batrick) [inline src/mds/MDCache.cc:9656]:
```suggestion
    auto& _mdr = active_requests[mdr->reqid];
```

**Review by batrick — CHANGES_REQUESTED** [review]
Let's move `mds: remove the stale class C_MDC_EmptyImport code ` to a separate PR. We won't want to backport that.

Otherwise LGTM

**Comment by ajarr** [comment]
> > The code looks good. I'm not sure about the commit message saying that it fixes https://tracker.ceph.com/issues/56633, which is the issue of duplicate/conflicting request ID.
> 
> Not a fix, just a test commit for this issue. Maybe `s/Fixes/URL/` ?

Not completely sure.

If you want to merge this commit and track its backporting, I think it would help to create another 'Cleanup' or a 'Bug' tracker in the CephFS project that wants to "Print details of conflicting requests that abort the MDS'.

If you don't want to merge this commit until you root cause/fix the conflicting request ID issue, then you can set this PR as draft.

There is also the "Partially-fixes" commit message tag. But I don't think this commit partially fixes https://tracker.ceph.com/issues/56633

Maybe others have better ideas. @vshankar ?

File: src/mds/MDCache.cc (hunk 2 above)
Comment (lxbsz) [reply src/mds/MDCache.cc]: Done. Thanks.

**Comment by lxbsz** [comment]
> Let's move `mds: remove the stale class C_MDC_EmptyImport code ` to a separate PR. We won't want to backport that.
> 

Done.

> > > The code looks good. I'm not sure about the commit message saying that it fixes https://tracker.ceph.com/issues/56633, which is the issue of duplicate/conflicting request ID.
> > 
> > 
> > Not a fix, just a test commit for this issue. Maybe `s/Fixes/URL/` ?
> 
> Not completely sure.
> 
> If you want to merge this commit and track its backporting, I think it would help to create another 'Cleanup' or a 'Bug' tracker in the CephFS project that wants to "Print details of conflicting requests that abort the MDS'.
> 
> If you don't want to merge this commit until you root cause/fix the conflicting request ID issue, then you can set this PR as draft, or add DNM PR label.
> 
> There is also the "Partially-fixes" commit message tag. But I don't think this commit partially fixes https://tracker.ceph.com/issues/56633
> 
> Maybe others have better ideas. @vshankar ?

I have created one child tracker for this PR.

Thanks!

**Review by batrick — APPROVED** [review]

**Comment by vshankar** [comment]
jenkins test make check

**Review by vshankar — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#2022-Aug-22

**Pull request merged**